File: src/main/java/com/smarthome/backend/server/api/modules/xiaomi/XiaomiVacuumCleaner.java
package com.smarthome.backend.server.api.modules.xiaomi;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.smarthome.backend.model.devices.DeviceVacuumCleaner;

/**
 * Reprasentiert einen Xiaomi Saugroboter als Geraet.
 * Erweitert {@link DeviceVacuumCleaner} und kapselt die Xiaomi-spezifische Kommunikation.
 */
public class XiaomiVacuumCleaner extends DeviceVacuumCleaner {
    private static final Logger logger = LoggerFactory.getLogger(XiaomiVacuumCleaner.class);

    /**
     * Netzadresse des Geraets (z. B. IP-Adresse).
     */
    private String address;

    /**
     * Token fuer die MiIO-Kommunikation.
     */
    private String token;

    /**
     * Modellbezeichnung des Geraets.
     */
    private String model;

    /**
     * Xiaomi Device-ID (DID), falls vorhanden.
     */
    private String did;

    /**
     * Standardkonstruktor (z. B. fuer Gson-Deserialisierung).
     */
    protected XiaomiVacuumCleaner() {
        super();
        this.address = null;
        this.token = null;
        this.model = null;
        this.did = null;
        super.isConnected = token != null;
        setModuleId("xiaomi");
    }

    /**
     * Konstruktor mit Pflichtfeldern.
     *
     * @param name Anzeigename des Geraets
     * @param id eindeutige Geraete-ID
     * @param address IP-Adresse oder Hostname
     * @param token MiIO-Token
     * @param model Modellbezeichnung
     * @param did Xiaomi Device-ID
     */
    public XiaomiVacuumCleaner(String name, String id, String address, String token, String model, String did) {
        super(name, id);
        this.address = address;
        this.token = token;
        this.model = model;
        this.did = did;
        super.isConnected = token != null;
        setModuleId("xiaomi");
        this.updateValues();
    }

    @Override
    public void updateValues() {
        if (this.address == null || this.token == null) {
            logger.debug("updateValues() uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("updateValues() nicht implementiert fuer {}", getId());
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getToken() {
        return token;
    }

    public void setToken(String token) {
        this.token = token;
        super.isConnected = token != null;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public String getDid() {
        return did;
    }

    public void setDid(String did) {
        this.did = did;
    }

    @Override
    protected void executeSetPower(Boolean power) {
        if (this.address == null || this.token == null) {
            logger.warn("executeSetPower uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeSetPower({}) nicht implementiert fuer {}", power, getId());
    }

    @Override
    protected void executeStartCleaning() {
        if (this.address == null || this.token == null) {
            logger.warn("executeStartCleaning uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeStartCleaning nicht implementiert fuer {}", getId());
    }

    @Override
    protected void executeStopCleaning() {
        if (this.address == null || this.token == null) {
            logger.warn("executeStopCleaning uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeStopCleaning nicht implementiert fuer {}", getId());
    }

    @Override
    protected void executePauseCleaning() {
        if (this.address == null || this.token == null) {
            logger.warn("executePauseCleaning uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executePauseCleaning nicht implementiert fuer {}", getId());
    }

    @Override
    protected void executeResumeCleaning() {
        if (this.address == null || this.token == null) {
            logger.warn("executeResumeCleaning uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeResumeCleaning nicht implementiert fuer {}", getId());
    }

    @Override
    protected void executeDock() {
        if (this.address == null || this.token == null) {
            logger.warn("executeDock uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeDock nicht implementiert fuer {}", getId());
    }

    @Override
    protected void executeSetMode(String mode) {
        if (this.address == null || this.token == null) {
            logger.warn("executeSetMode uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeSetMode({}) nicht implementiert fuer {}", mode, getId());
    }

    @Override
    protected void executeCleanRoom(String roomId) {
        if (this.address == null || this.token == null) {
            logger.warn("executeCleanRoom uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeCleanRoom({}) nicht implementiert fuer {}", roomId, getId());
    }

    @Override
    protected void executeCleanZone(String zoneId) {
        if (this.address == null || this.token == null) {
            logger.warn("executeCleanZone uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeCleanZone({}) nicht implementiert fuer {}", zoneId, getId());
    }

    @Override
    protected void executeChangeFanSpeed(int fanSpeed) {
        if (this.address == null || this.token == null) {
            logger.warn("executeChangeFanSpeed uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeChangeFanSpeed({}) nicht implementiert fuer {}", fanSpeed, getId());
    }

    @Override
    protected void executeChangeMode(String mode) {
        if (this.address == null || this.token == null) {
            logger.warn("executeChangeMode uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeChangeMode({}) nicht implementiert fuer {}", mode, getId());
    }

    @Override
    protected void executeClearError() {
        if (this.address == null || this.token == null) {
            logger.warn("executeClearError uebersprungen fuer {} - address/token fehlen", getId());
            return;
        }
        logger.debug("executeClearError nicht implementiert fuer {}", getId());
    }
}

File: src/main/node/com/smarthome/backend/server/api/modules/xiaomi/xiaomiVacuumCleaner.ts
import { logger } from "../../../../logger.js";
import { DeviceVacuumCleaner } from "../../../../model/devices/DeviceVacuumCleaner.js";
import { XiaomiController } from "./xiaomiController.js";

export class XiaomiVacuumCleaner extends DeviceVacuumCleaner {
  private address?: string;
  private token?: string;
  private model?: string;
  private did?: string;

  constructor(
    name?: string,
    id?: string,
    address?: string,
    token?: string,
    model?: string,
    did?: string
  ) {
    super();
    if (name) this.name = name;
    if (id) this.id = id;
    this.address = address;
    this.token = token;
    this.model = model;
    this.did = did;
    this.isConnected = Boolean(token);
    this.moduleId = "xiaomi";
    this.updateValues();
  }

  updateValues() {
    if (!this.address || !this.token) {
      logger.debug("updateValues() uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    logger.debug("updateValues() nicht implementiert fuer {}", this.id);
  }

  getAddress() {
    return this.address;
  }

  setAddress(address: string) {
    this.address = address;
  }

  getToken() {
    return this.token;
  }

  setToken(token: string) {
    this.token = token;
    this.isConnected = Boolean(token);
  }

  getModel() {
    return this.model;
  }

  setModel(model: string) {
    this.model = model;
  }

  getDid() {
    return this.did;
  }

  setDid(did: string) {
    this.did = did;
  }

  protected executeSetPower(power: boolean) {
    if (!this.address || !this.token) {
      logger.warn("executeSetPower uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "setPower", power);
  }

  protected executeStartCleaning() {
    if (!this.address || !this.token) {
      logger.warn("executeStartCleaning uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "start");
  }

  protected executeStopCleaning() {
    if (!this.address || !this.token) {
      logger.warn("executeStopCleaning uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "stop");
  }

  protected executePauseCleaning() {
    if (!this.address || !this.token) {
      logger.warn("executePauseCleaning uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "pause");
  }

  protected executeResumeCleaning() {
    if (!this.address || !this.token) {
      logger.warn("executeResumeCleaning uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "resume");
  }

  protected executeDock() {
    if (!this.address || !this.token) {
      logger.warn("executeDock uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "home");
  }

  protected executeSetMode(mode: string) {
    if (!this.address || !this.token) {
      logger.warn("executeSetMode uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "setMode", mode);
  }

  protected executeCleanRoom(roomId: string) {
    if (!this.address || !this.token) {
      logger.warn("executeCleanRoom uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "cleanRoom", roomId);
  }

  protected executeCleanZone(zoneId: string) {
    if (!this.address || !this.token) {
      logger.warn("executeCleanZone uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "cleanZone", zoneId);
  }

  protected executeChangeFanSpeed(fanSpeed: number) {
    if (!this.address || !this.token) {
      logger.warn("executeChangeFanSpeed uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "setFanSpeed", fanSpeed);
  }

  protected executeChangeMode(mode: string) {
    if (!this.address || !this.token) {
      logger.warn("executeChangeMode uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "setMode", mode);
  }

  protected executeClearError() {
    if (!this.address || !this.token) {
      logger.warn("executeClearError uebersprungen fuer {} - address/token fehlen", this.id);
      return;
    }
    void XiaomiController.callMethod(this.address, this.token, "reset");
  }
}

